source/h2_stream.c

```diff
@@ -586,6 +588,46 @@ static int s_stream_reset_stream(struct aws_http_stream *stream_base, uint32_t h
     return s_stream_reset_stream_internal(stream_base, stream_error);
 }
 
+void s_stream_cancel(struct aws_http_stream *stream_base, int error_code) {
+    struct aws_h2err stream_error = {
+        .aws_code = error_code,
+        .h2_code = AWS_HTTP2_ERR_CANCEL,
+    };
+
+    struct aws_h2_stream *stream = AWS_CONTAINER_OF(stream_base, struct aws_h2_stream, base);
+    struct aws_h2_connection *connection = s_get_h2_connection(stream);
+    bool reset_called;
+    bool stream_is_active;
+    bool cross_thread_work_should_schedule = false;
+
+    { /* BEGIN CRITICAL SECTION */
+        s_lock_synced_data(stream);
+
+        reset_called = stream->synced_data.reset_called;
+        stream_is_active = stream->synced_data.api_state == AWS_H2_STREAM_API_STATE_ACTIVE;
+        if (!reset_called && stream_is_active) {
+            cross_thread_work_should_schedule = !stream->synced_data.is_cross_thread_work_task_scheduled;
+            stream->synced_data.reset_called = true;
```

Comment:
this looks a lot like the implementation of s_stream_reset_stream_internal(). Why not just call that? If there's some difference, can we just modify that function? it's internal